packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+   However, if the client is a read-only client, then it can't do this, and it needs to retain the code to read view schema A in case it opens a document with stored schema A.
+   This caveat is made more manageable by [adapters](#adapters), introduced later in this document.
+
+     > To minimize the time between client saturation and document upgrades, the ability to upgrade the document can be made part of the deployment in step 1 but "shipped dark" and remotely enabled in this step via a feature flight.
+
+4. Wait until all documents have upgraded to the new stored schema.
+   After this **document saturation** has been achieved, the application can do another deployment, removing view schema A and all of its related code, if it was not able to do so in step 3 (because some of its clients are read-only).
+   It's important that this happens after document saturation, or else there will be read-only clients that fail to open a document with stored schema A because they only know about view schema B.
+
+   > In some environments, there's no guarantee that document saturation will ever occur - and this step is never completed.
+     Old documents can hang around forever if they remain unopened, but must still be supported in the event that someday they are needed.
+     The code that reads view schema A must be maintained forever.
+     Fortunately, this headache can be somewhat mitigated via the use of [adapters](#adapters).
+
+5. Eventually, all clients will have updated to the latest deployment, and all traces of schema A will be gone in both the documents and application source code.
+
+```mermaid
+gantt
+dateFormat D
+axisFormat Step %e
+tickInterval 1day
+Maintain view schema A in source code (read/write client) : 1, 2d
+Maintain view schema A in source code (read-only client)  : 1, 3d
+Documents contain stored schema A                         : 1, 3d
+Maintain view schema B in source code                     : 2, 3d
+Documents contain stored schema B                         : 3, 2d
+```
+
+Different applications with different requirements may tweak this process in various ways, but in general doing a staged rollout is a cumbersome endeavor.
+It takes time, it requires code bifurcation/duplication, and it's complicated to understand and coordinate.
+The next sections will explore ways to reduce the amount of effort a developer must take to change their schema.
+
+### Adapters
+
+A **schema adapter** helps to reduce the code bifurcation/duplication (described in the previous section) that results from doing a staged rollout.
+One of the main pain points of the staged rollout process is that the application has to maintain two view schema while the migration is ongoing (or forever!), since the client might encounter documents of either the old stored schema or the new stored schema.
+In practice, this can mean a lot of redundant application code as the UI is bound to two different data models, but only one is ever used at a time.
```

Comment:
Consider not using UI, instead find something neutral (application may have no UI)